The small workaround fixes the missing MD5 for Jottacloud by using a …

olihey opened the pull request:
…local cache directory

This is a fix for the required MD5 before upload to Jottacloud.
If a source can't provide a MD5 (like the crypt backend for example) the entire source file is read into a file in a local cache directory and the uploaded after the full file is read including MD5 calculation.
The local cache directory needs to be define during `config` of the Jottacloud remote

ncw commented:
I suggest you take a look at the code I deleted here b510c70c1ef2bd41eee962dc04fbbe0880edebd6 which is a battle tested sha1sum program.  Note it uses go's tempfile facility which you can redirect wherever you want with the TMPDIR environment variable.

olihey commented:
Thanks for the pointer, not sure I am liking the TMPDIR solution as one might forgets this and ends up filling up it's root (C:\ drive). This will not be happening if a user can specify the directory themselves during `config`. Although having a fixed directory for the remote might "restrain" the user.

I'll look into it

buengese commented:
I also just started implementing this but as you already opened a pull request you may as well do this. Another improvement I would have implemented is to cache small files e.g < 10mb in ram rather than create a tempfile for them as this is unnecessary wear on the hard drive.

olihey commented:
I removed the localDir option and use the `ioutil.TempFile` as suggested by @ncw and also added a threshold for NOT caching files on disk if the smaller than this limit (default 10M) as suggested by @buengese

vb0 commented:
Works for me local->crypt/jotta.
Both small and large files (I set to 1GB the limit and I have some above).

buengese commented:
Fix looks fine to me now. I would be in favor of merging this as soon as possible. Just maybe squash it down to 1 or 2 commits.

olihey commented:
Small note, I saw that the `ioutil.Tempfile` leaves the files behind when you kill rclone using Ctrl-C. That's actually quite bad but I haven't found a way to fix that.

mpolden commented:
> Small note, I saw that the ioutil.Tempfile leaves the files behind when you kill rclone using Ctrl-C. That's actually quite bad but I haven't found a way to fix that.

Typically, you would use `signal.Notify` to trigger clean up for given signals (`SIGINT`, and likely `SIGTERM`) and clean up any temporary files. Not sure if rclone installs a signal handler that backends can hook into, or if it's up to each backend to do the appropriate thing.

vb0 commented:
I've noticed that too, might be particularly bad as the highest chance to interrupt rclone is during very long transfers of very large files. I guess this is usually handled by catching SIGTERM and cleaning up (I don't know how it's done in go and if something special needs to be done in rclone to support these manipulations). Still won't help for SIGKILL (or crashes) ... in this case a cleanup task would be needed ... which would have to take care if other rclone processes are running ... but then if there are other processes running it doesn't mean they actually have the files opened and we need a mechanism to see if the existing files are in use by processes which are alive and kicking ...

Probably not worth it at this point, who knows what changes we'll have next week - I've got this mandatory md5 thing literally the first day I signed up with them...

buengese commented:
Rclone allready handles signals see https://github.com/ncw/rclone/blob/6791cf7d7f06de838cafecbe63892ff3ddb90ceb/lib/atexit/atexit.go#L22-L42

mpolden commented:
Great! Registering a clean up function should be simple enough then.

olihey commented on backend/jottacloud/jottacloud.go:
```diff
@@ -748,6 +759,59 @@ func (o *Object) Open(options ...fs.OpenOption) (in io.ReadCloser, err error) {
+			}
+
+			// clean up the file after we are done downloading
+			defer func() {
```
You are absolutely right, fix commited

olihey replied:
```
cleanup := func() {
  _ = tempFile.Close()
  _ os.Remove(tempFile.Name())
}
defer cleanup()
atexit.Register(cleanup)
```
Wouldn't that just create a memory leak for long running tasks because we collecting and keeping all tempfile objects up until rclone exits.
There should be just one cleanup function the removes the files by using a list of open tempfiles and closing and removing them.

mpolden replied:
Good point. How about avoiding capturing `tempFile` in the cleanup function that is passed to `atexit.Register`? 

```
tempName := tempFile.Name()
atexit.Register(func() { _ = os.Remove(tempName) })
```

The OS will clean up any open file descriptors when the program terminates anyway.

ncw replied:
In the code I posted, it deletes the file after opening.  On all OSes except Windows this will mean that we use a temporary anonymous file which will get removed when it is closed or the program exits unexpectedly.

https://github.com/ncw/rclone/commit/b510c70c1ef2bd41eee962dc04fbbe0880edebd6#diff-9d62ca00aab4a67defb33a6249776bc3L1263

For the Windows case - the temporary file is in the users TEMP directory which there are friendly Windows way of cleaning.

Using atexit would get some of the additional cases (but not all) that can happen when rclone exits so I'm not sure it is worth it.

olihey commented:
If I add
```
// be kind and clean up
atexit.Register(func() {
	fmt.Println("YES")
})
```
at the end of `NewFs` in the `jottacloud.go` file `Ctrl-C` does not work at all on my Windows machine anymore. Rclone just keeps running no matter how often I press Ctrl-C

I am a bit stuck here and won't have time to look at it until end of next week, sorry!

ncw commented on backend/jottacloud/jottacloud.go:
```diff
@@ -748,6 +759,59 @@ func (o *Object) Open(options ...fs.OpenOption) (in io.ReadCloser, err error) {
 func (o *Object) Update(in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (err error) {
+	md5String, err := src.Hash(hash.MD5)
+	if err != nil || md5String == "" {
+		// if the source can't provide a MD5 hash we are screwed and need to calculate it ourselfs
```
This little bit of code is now complicated enough that I'd factor it out into a function

```go
func readMD5(in io.Reader, size, threshold int64) (string, io.Reader, error) {}
```

and write a unit test that exercises all the paths in it.  The integration tests will only test the < 10M path.

olihey replied:
Awesome, thanks

ncw commented:
@olihey wrote:

> If I add
> 
> // be kind and clean up
> atexit.Register(func() {
> 	fmt.Println("YES")
> })
> at the end of NewFs in the jottacloud.go file Ctrl-C does not work at all on my Windows machine anymore. Rclone just keeps running no matter how often I press Ctrl-C
> 
> 
> at the end of NewFs in the jottacloud.go file Ctrl-C does not work at all on my Windows machine anymore. Rclone just keeps running no matter how often I press Ctrl-C

That sounds like a bug!

> I am a bit stuck here and won't have time to look at it until end of next week, sorry!

If you want me to finish off the last few things on this PR then I will if you haven't got time.

olihey commented:
@ncw: Would be great if you can find the time.

mpolden commented:
Was going to volunteer to do the fixes too, but also realized I didn't have time until this week. 

Great that you're on it, @ncw. Let me know if you need some help, e.g. to review or test.

ncw commented:
I've merged that into a branch and made a beta from it here: fb53e137c07f551cf08e5646fc22bd91bb10bb95 - I added tests, fixes and docs on top here 86a21af40d7a496247f85520fc1229754918d1ac

This is in

https://beta.rclone.org/branch/v1.42-120-g86a21af4-fix-2462-jottacloud-md5/ (uploaded in 15-30 mins)

Thank you @olihey for fixing this :-)

mpolden commented:
Seems to work for me.

Uploaded/downloaded around 10GB of data which made the expected temp files. Accounting of progress looked correct too.
